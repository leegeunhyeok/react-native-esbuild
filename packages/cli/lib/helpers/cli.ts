import { colors } from '@react-native-esbuild/shared';
import { logger } from '../shared';

export const getCommand = <RawArgv extends { _: (string | number)[] }>(
  argv: RawArgv,
  position = 0,
): string => argv._[position].toString();

// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Allow any type for debugging.
export const printDebugOptions = <T extends Record<string, any>>(
  options: T,
): void => {
  Object.entries(options).forEach(([key, value], index, entries) => {
    const isLast = entries.length - 1 === index;
    const pipe = `${isLast ? '╰' : '├'}─`;
    const keyValue = `${key}: ${JSON.stringify(value)}`;
    logger.debug(colors.gray(`${pipe} ${keyValue}${isLast ? '\n' : ''}`));
  });
};
