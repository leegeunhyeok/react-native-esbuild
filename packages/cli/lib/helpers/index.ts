export * from './cli';
export * from './core';
export * from './errors';
export * from './stdin';
