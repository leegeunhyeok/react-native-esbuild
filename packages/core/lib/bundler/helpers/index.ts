export * from './async';
export * from './config';
export * from './fs';
export * from './internal';
