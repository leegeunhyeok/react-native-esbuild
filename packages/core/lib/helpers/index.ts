export * from './async';
export * from './internal';
