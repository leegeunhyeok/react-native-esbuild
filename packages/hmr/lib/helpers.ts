import path from 'node:path';
import {
  colors,
  type BuildContext,
  type ModuleId,
} from '@react-native-esbuild/shared';
import type { Metafile } from 'esbuild';
import { getReloadByDevSettingsProxy } from '@react-native-esbuild/internal';
import { HMR_REGISTER_FUNCTION, HMR_UPDATE_FUNCTION } from './constants';
import { logger } from './shared';

let INTERNAL_boundaryIndex = 0;

export const asHMRBoundary = (id: ModuleId, body: string): string => {
  const ident = `__hmr${INTERNAL_boundaryIndex++}`;
  return `var ${ident} = ${HMR_REGISTER_FUNCTION}(${JSON.stringify(id)});
  ${body}
  ${ident}.dispose(function () { });
  ${ident}.accept(function (payload) {
    global.__hmr.reactRefresh.performReactRefresh();
  });`;
};

/**
 * Make code that wrap body with HMR API's update callback.
 *
 * **example**
 * ```js
 * global.__hmr.update(id, function () {
 *   {body}
 * });
 * ```
 */
export const asHMRUpdateCall = (id: ModuleId, body: string): string => {
  return `${HMR_UPDATE_FUNCTION}(${JSON.stringify(id)}, function () {
    ${body}
  });`;
};

/**
 * Make code that wrap with try-catch block with fallback handler
 *
 * **example**
 * ```js
 * try {
 *   {body}
 * } catch (error) {
 *   console.error('[HMR] unable to accept', error);
 *   reload();
 * }
 * ```
 */
export const asFallbackBoundary = (body: string): string => {
  return `try {
    ${body}
  } catch (error) {
    console.error('[HMR] unable to accept', error);
    ${getReloadByDevSettingsProxy()}
  }`;
};

/**
 * Make code that register as external module.
 *
 * **example**
 * ```js
 * {body}
 * global.__modules.external(id, identName);
 * ```
 */
export const registerAsExternalModule = (
  id: ModuleId,
  body: string,
  identifierName: string,
): string => {
  return `${body}\nglobal.__modules.external(${JSON.stringify(
    id,
  )}, ${identifierName});`;
};

export const isHMRBoundary = (path: string): boolean => {
  return !path.includes('/node_modules/') && !path.endsWith('runtime.js');
};

/**
 * Get actual imported module path.
 */
export const getActualImportPaths = (
  buildContext: BuildContext,
  imports: Metafile['inputs'][string]['imports'],
): Record<string, string> => {
  const importPaths = imports.reduce(
    (prev, curr) => {
      // To avoid wrong assets path.
      // eg. `react-native-esbuild-assets:/path/to/assets`
      const splitted = path.resolve(buildContext.root, curr.path).split(':');
      const actualPath = splitted[splitted.length - 1];
      if (curr.original && !prev[curr.original]) {
        logger.debug(
          `${colors.gray(
            `├─ ${stripRoot(buildContext.root, curr.original)} ▸`,
          )} ${
            isExternal(
              buildContext.additionalData.externalPattern as string,
              curr.original,
            )
              ? colors.gray('<external>')
              : stripRoot(buildContext.root, actualPath)
          }`,
        );
      }
      return {
        ...prev,
        ...(curr.original ? { [curr.original]: actualPath } : null),
      };
    },
    {} as Record<string, string>,
  );

  logger.debug(colors.gray(`╰─ ${Object.keys(importPaths).length} import(s)`));

  return importPaths;
};

const stripRoot = (rootPath: string, path: string): string =>
  path.replace(new RegExp(`^${rootPath}/?`), '');

const isExternal = (pattern: string, path: string): boolean =>
  new RegExp(pattern).test(path);
