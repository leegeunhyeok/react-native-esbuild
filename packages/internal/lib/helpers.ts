import path from 'node:path';
import indent from 'indent-string';

export const resolveFromRoot = (
  targetPath: string,
  root = process.cwd(),
): string => {
  return require.resolve(targetPath, {
    paths: Array.from(
      new Set([
        // add current workspace directory to module resolution path.
        // improve for workspace projects (eg. monorepo)
        path.join(root, 'node_modules'),
        ...(require.main?.paths ?? []),
      ]),
    ).filter(Boolean),
  });
};

export const wrapWithIIFE = (body: string, filepath: string): string => `
// ${filepath}
(function (global) {
${indent(body, 2)}
})(typeof globalThis !== 'undefined' ? globalThis : typeof global !== 'undefined' ? global : typeof window !== 'undefined' ? window : this);
`;
