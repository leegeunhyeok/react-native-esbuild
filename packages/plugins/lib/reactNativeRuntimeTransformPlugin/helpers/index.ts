export * from './caches';
export * from './json';
