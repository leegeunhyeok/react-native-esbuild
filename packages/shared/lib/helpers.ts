import path from 'node:path';
import md5 from 'md5';
import { OptionFlag } from './enums';
import type { BundleOptions, Id } from './types/core';
import {
  DEFAULT_ENTRY_POINT,
  DEFAULT_OUTFILE,
  LOCAL_CACHE_DIR,
  PUBLIC_PATH,
  STATUS_CACHE_FILE,
} from './constants';

export const combineWithDefaultBundleOptions = (
  options: Partial<BundleOptions>,
): BundleOptions => {
  if (!options.platform) {
    throw new Error('platform is required');
  }

  return {
    ...options,
    platform: options.platform,
    minify: options.minify ?? false,
    dev: options.dev ?? true,
    entry: path.resolve(options.entry ?? DEFAULT_ENTRY_POINT),
    outfile: path.resolve(options.outfile ?? DEFAULT_OUTFILE),
    metafile: options.metafile ?? false,
  };
};

export const getIdByOptions = ({
  platform,
  dev,
  minify,
}: Pick<BundleOptions, 'platform' | 'dev' | 'minify'>): number => {
  let value = OptionFlag.None; // = 0

  // platform
  value |= platform === 'android' ? OptionFlag.PlatformAndroid : 0;
  value |= platform === 'ios' ? OptionFlag.PlatformIos : 0;
  value |= platform === 'web' ? OptionFlag.PlatformWeb : 0;

  // dev & minify
  value |= dev ? OptionFlag.Dev : 0;
  value |= minify ? OptionFlag.Minify : 0;

  return value;
};

/**
 * Generate hash that contains the file path, modification time, and bundle options.
 *
 * `id` is combined(platform, dev, minify) bundle options value in `@react-native-esbuild`.
 *
 * hash = md5(id + modified time + file path + core version)
 *            number + number    + string    + string
 *
 * @param id - id
 * @param filepath - file path
 * @param modifiedAt - file modified at timestamp
 * @returns
 */
export const getCacheKey = (
  id: Id,
  filepath: string,
  modifiedAt: number,
): string => {
  return md5(id + modifiedAt + filepath + (self._version as string));
};

/**
 * For resolve environment mismatch issue.
 *
 * Run bundler with `--dev=false` (production), but babel's env is still using `development` environment.
 * In this case, run bundle after build it will be occurs unexpected errors.
 *
 * For eg. `react-native-reanimated` using babel plugin for transform.
 *         Plugin follows babel environment, but main source(by esbuild) isn't.
 *
 * Override `NODE_ENV`, `BABEL_ENV` to bundler's environment.
 *
 * @param isDev - env is 'development'
 * @see {@link https://github.com/babel/babel/blob/v7.23.0/packages/babel-core/src/config/helpers/environment.ts#L2}
 */
export const setEnvironment = (isDev: boolean): void => {
  const env = isDev ? 'development' : 'production';
  process.env.NODE_ENV = env;
  process.env.BABEL_ENV = env;
};

export const getDevServerPublicPath = (root: string): string => {
  return path.resolve(root, LOCAL_CACHE_DIR, PUBLIC_PATH);
};

export const getBuildStatusCachePath = (root: string): string => {
  return path.resolve(root, LOCAL_CACHE_DIR, STATUS_CACHE_FILE);
};
