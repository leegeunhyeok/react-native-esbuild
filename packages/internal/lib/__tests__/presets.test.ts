import { getAssetRegistrationScript } from '../helpers';

describe('getAssetRegistrationScript', () => {
  let assetRegistrationScript: string;

  beforeEach(() => {
    assetRegistrationScript = getAssetRegistrationScript({
      name: 'image',
      type: 'png',
      scales: [1, 2, 3],
      hash: 'hash',
      httpServerLocation: '/image.png',
      dimensions: {
        width: 0,
        height: 0,
      },
    });
  });

  it('should match snapshot', () => {
    expect(assetRegistrationScript).toMatchSnapshot();
  });
});
