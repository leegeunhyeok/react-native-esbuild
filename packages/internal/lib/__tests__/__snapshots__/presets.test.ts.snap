// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAssetRegistrationScript should match snapshot 1`] = `"module.exports =require('react-native/Libraries/Image/AssetRegistry').registerAsset({"__packager_asset":true,"name":"image","type":"png","scales":[1,2,3],"hash":"hash","httpServerLocation":"/image.png","width":0,"height":0});"`;
